vscode/src/telemetry.ts

```diff
@@ -0,0 +1,77 @@
+import * as vscode from "vscode";
+import TelemetryReporter, {
+  TelemetryEventMeasurements,
+  TelemetryEventProperties,
+} from "@vscode/extension-telemetry";
+import { log } from "qsharp-lang";
+
+export enum EventType {
+  DebugSessionStart = "DebugSessionStart",
+  Initialize = "Initialize",
+  LoadLanguageService = "LoadLanguageService",
+  QSharpJupyterCellInitialized = "QSharpJupyterCellInitialized",
+}
+
+type Empty = { [K in any]: never };
+
+type EventTypes = {
+  [EventType.Initialize]: {
+    properties: Empty;
+    measurements: Empty;
+  };
+  [EventType.DebugSessionStart]: {
+    properties: Empty;
+    measurements: {
+      timeToStartMs: number;
+    };
+  };
+  [EventType.LoadLanguageService]: {
+    properties: Empty;
+    measurements: {
+      timeToStartMs: number;
+    };
+  };
+  [EventType.QSharpJupyterCellInitialized]: {
+    properties: Empty;
+    measurements: Empty;
+  };
+};
+
+type WrappedTelemetryEvent = {
+  id: string;
+  data?: {
+    measurements: TelemetryEventMeasurements;
+    properties: TelemetryEventProperties;
+  };
+};
+
+export function initTelemetry() {
+  const extension = vscode.extensions.getExtension("quantum.qsharp-vscode");
+  const packageJson = extension?.packageJSON;
+  if (!packageJson) {
+    return;
+  }
+  // see issue here: https://github.com/microsoft/vscode-extension-telemetry/issues/183
+  // we cannot use the latest version of extension-telemetry until this is fixed
+  const reporter = new TelemetryReporter(
+    "qsharp-vscode",
+    packageJson.version,
+    packageJson.aiKey
+  );
+  log.setTelemetryCollector(
+    ({
+      id,
+      data: { properties, measurements } = { properties: {}, measurements: {} },
+    }: WrappedTelemetryEvent) =>
+      reporter.sendTelemetryEvent(id, properties, measurements)
+  );
+  sendTelemetryEvent(EventType.Initialize, {}, {});
+}
+
+export function sendTelemetryEvent<E extends keyof EventTypes>(
```

Comment:
The intent was that code, be it in npm, playground, or VS Code, would call log.logTelemetry to send events, and then whatever telemetry consumer had been wired-up would be hooked into that. Seems odd here that we're exposing a sendTelemetryEvent API for use in VS Code, and then that calls log.logTelemetry. Can we not just wire up so the VS Code source also calls log.logTelemetry? Happy to refactor that API if you think it's too loose.

Reply:
> Since, if I'm not mistaken, the VSCode extension is the only consumer of the API

You are mistaken :-)  A search of the playground code will show one place where we call it (mostly to prove it worked, happy for you to refactor that also)